proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,858 @@
+
+An exception to this rule is ``newtype`` declarations in GADT syntax:
+``newtype``-s must have only linear arguments (see Interactions
+below). For backward compatibility, we propose, to make unrestricted arrow
+``(->)`` in ``newtype``-s be interpreted as linear arrows, and create
+a new warning ``unrestricted-newtype`` triggered when this happens.
+
+Base
+~~~~
+
+Because linear functions only strengthen the contract of unrestricted
+functions, a number of functions of ``base`` can get a more precise
+type. However, for pedagogical reason, to prevent linear types from
+interfering with newcomers' understanding the ``Prelude``, this
+proposal does not modify ``base``. Instead we will release a library
+exposing the stronger types for ``base`` functions.
+
+This library will not redefine any type, and instead takes advantage
+of the fact that data types, in ``base`` are linear by default to
+reuse the same types, hence remain compatible with base.
+
+The only function which will need to change is ``($)`` because its
+typing rule is built in the type checker. Ignoring the details about
+levity and higher-rank polymorphism in the typing rule, the type
+``($)`` will be:
+
+::
+
+  ($) :: (a :p-> b) ⊸ a :p-> b
+
+The precise content of the library is out of scope of this proposal:
+future standardisation of library content is the competence of
+the CLC.  However the library will also contain convenient types to
+work with linear types, with the understanding that when the new types
+are standardised in ``base`` the library would re-export them rather
+than define them, such as:
+
+::
+
+   data Unrestricted a where
+     Unrestricted :: a -> Unrestricted a
+
+Formalism
+~~~~~~~~~
+
+So far, we have considered only two multiplicities ``1`` and
+``ω``. But the metatheory works with any so-called
+sup-semi-lattice-ordered semi-ring of multiplicities. That is: there
+is a 0, a 1, a sum and a product with the usual distributivity laws, a
+(computable) order compatible with the sum and product, such that each
+pair of multiplicities has a (computable) join. Even if there is only
+three multiplicities in this proposal, the proposal is structured to
+allow future extensions.
+
+Non of our examples use ``0``, however, ``0`` turns out to be useful
+for implementation of type-checking. Additionally, ``0`` has been used
+by `Conor McBride
+<https://link.springer.com/chapter/10.1007/978-3-319-30936-1_12>`_ to
+handle dependent types, which may matter for Dependent Haskell). In
+both case, the use of ``0`` could be seen as an internal use, but
+there is no real reason to deny access to the programmer. Hence it is
+included in the syntax.
+
+Here is the definition of sum, product and order for this proposal's
+multiplicities (in Haskell pseudo-syntax):
+
+::
+   0 + x = x
+   x + 0 = x
+   _ + _ = ω
+
+   0 * _ = 0
+   _ * 0 = 0
+   1 * x = x
+   x * 1 = 1
+   ω * ω = ω
+
+   _ ⩽ ω = True
+   x ⩽ y = x == y
+
```

Comment:
Attention! in particular 0 <= 1 is NOT true! (Multiplicities should be understood as *sets* of possible usages and <= as set inclusion)

Reply:
Done.